### ISSUE TITLE
Create Umbraco Package (and nuget) that will install it all nicely

### ISSUE THREAD

**Issue by PeteDuncanson** [rasmusjp/umbraco-graphql#25]
I'm trying to get this packaged up in time for the USA Festival in a few weeks. To get a MVP package out we (@offroadcode) are working on a UI to manage access to individual fields that takes the form of a dashboard in the back office. Currently this is a separate repos (as we just cloned are base Package repos) and not sure if we should merge it in with this one or keep this as sort of the Core of the code (@rasmusjp any preference?). 

See the repos here https://github.com/Offroadcode/GraphQL-for-Umbraco

Current aim is to have it use a Default user which will have access to zero fields out of the box. Site admins (well anyone with dev section access actually) will have to whitelist which fields the Default user can see. These fields will then be accessible publicly via the GraphQL endpoint and GraphiQL for the site. As 95% of sites will just want to power their frontend with this for now I think its totally ok to just roll out with Default user only and then we can add in the other user management and access token stuff later (ideally once I've had chance to talk this through over a beer with folks face to face at USA Fest, Denmark Fest and UK Fest) and gathered some thoughts/ideas.

So in short the MVP will need:
- ~~Basic UI to manage which fields to show per doctype~~
- Back Office API's to save the selections from the UI
- ~~Ship with a default user~~
- ~~Create all the tables needed on start up as per #10~~
- Packaged up as an Umbraco Package
- Ideally available as a Nuget Package too

**Comment by rasmusjp** [comment]
The link’s giving me a 404, maybe it’s private?

I think it would be nice to have the code at the same place for discoverability, but it might just be different repos in the same organisation? (I’ve registered https://github.com/GraphQL-for-Umbraco but I’m open for suggestions if we go this route)

If we do merge it with this repo, I still think it should be released as separate packages, so we have a Core package, and X number of “addon” packages.

I personally think that a NuGet package is a must have these days and will happily set up a build script for it if needed.

One thing we need to figure out is how we do the [Owin registration](
https://github.com/rasmusjp/umbraco-graphql/blob/1284369fd937a0081d633bf39419a4bc780c9d58/samples/Website/App_Start/UmbracoEvents.cs#L12-L16), do we do it automatically since there’s really no options right now or do the implementer need to do some configuration (which I think would be fine)?

Also before release we need to update the readme with the updated queries, since the examples in the readme right now won’t work.

**Comment by PeteDuncanson** [comment]
Ahh yes, it is. Should be public now :)

Owin is an issue, did you have some funky hook you could use that we listed somewhere? https://issues.umbraco.org/issue/U4-11169

Found this btw but its a bit old https://github.com/umbraco/UmbracoIdentityExtensions/blob/master/src/Umbraco.IdentityExtensions/App_Start/UmbracoStandardOwinStartup.cs.pp referenced from here https://our.umbraco.com/forum/umbraco-7/using-umbraco-7/63725-OWIN-Startup#comment-243811

Would be great to have this in Nuget before the talk and updated examples. Right now I'm full on with writing my slides and wrapping up business before I fly on Saturday :)

**Comment by rasmusjp** [comment]
I've created a PR in #26 with a build script for creating the NuGet package.

**Comment by PeteDuncanson** [comment]
Not happy with the name of our Package, it causes confusion. Can't think of a better name for it though. Open to suggestions.

**Comment by rasmusjp** [comment]
With #29 merged in I think it makes sense to also merge in the UI for it, since there's no (easy) way to manage the permissions otherwise.

**Comment by rasmusjp** [comment]
I've got the UI to build over in #31 and included in the NuGet package.

One thing I've noticed when testing it out was that I was not able to query for built in fields, e.g. when doing 

```graphql
{
  content {
    atRoot {
      all {
        _contentData {
          id
        }
      }
    }
  }
}
```
I'm just getting the auth required error back. Am I missing something or is it not possible to specify permissions for the built in fields right now?

**Comment by JackPenney** [comment]
Great :)

Currently the UI doesn't have the ability to do that yet, although the API its hitting does have the ability to set permissions for the built in fields on a per doctype basis

There's a few things to iron out relating to that and some decisions to made though, i'll spin up separate issues for them to get a discussion going and to keep track of what needs sorting

**Comment by rasmusjp** [comment]
Okay, I think it's something that's needed in the UI before we can do a release, not sure how easy it would be to add though.

We could always do a release without the authorisation and then look into how we get it working for the next release.

**Comment by JackPenney** [comment]
I've made an update to the ui so we can set build in properties for each doctype in this PR https://github.com/rasmusjp/umbraco-graphql/pull/33

Although i've found a snag where if you request content of the type PublishedContent the doctype permissions don't apply to what's requested, so that will still need sorting still. I've raised on issue for it here:
https://github.com/rasmusjp/umbraco-graphql/issues/32

If we need to release without auth for now then that's fine, in the meantime i'll take a look into https://github.com/rasmusjp/umbraco-graphql/issues/32 and see what i can do

**Comment by rasmusjp** [comment]
I'd love to do a release with auth in, but with #32 and #34, it'll need some more work and testing before I'm comfortable releasing it.

So if you think it's fine, I'll see if I can get it out on NuGet today or tomorrow.

**Comment by rasmusjp** [comment]
I've updated the examples and added installation instructions over in the [release-preparation](/rasmusjp/umbraco-graphql/tree/release-preparation) branch. It's created from the commit before #29 was merged in.

Let me know if I should release a NuGet package based on that branch.

**Comment by JackPenney** [comment]
Yeah i think a NuGet package based on that branch would be good

**Comment by rasmusjp** [comment]
Great, I've uploaded the package on [Nuget](https://www.nuget.org/packages/Our.Umbraco.GraphQL), it should be public soon

**Comment by PeteDuncanson** [comment]
Just a reminder the UI was a super quick hack job as a proof of concept. More work required and happy to do it :)

**Comment by rasmusjp** [comment]
Hi

Just wanted to let you know that the project is moving to the [Umbraco Community](https://github.com/umbraco-community/) GitHub organisation, so we are closing all existing issues.

If you think your issue is still relevant, please feel free to reopen it.

/Rasmus
